app/models/package_manager/go.rb

```diff
@@ -299,5 +267,36 @@ def self.canonical_module_name(name)
     def self.encode_for_proxy(str)
       str.gsub(/[A-Z]/) { |s| "!#{s.downcase}" }
     end
+
+    private_class_method def self.latest_version_number(name)
+      json = get_json("#{PROXY_BASE_URL}/#{encode_for_proxy(name)}/@latest")
+      number = json&.dig("Version")
+
+      Rails.logger.info "[Unable to fetch latest version number] name=#{name}" if number.blank?
+
+      number
+    end
+
+    private_class_method def self.fetch_mod(name, version: nil)
+      mod_contents = fetch_mod_contents(name, version: version)
+      return unless mod_contents
+
+      GoMod.new(mod_contents)
+    end
+
+    private_class_method def self.fetch_mod_contents(name, version: nil)
+      # Go proxy spec: https://golang.org/cmd/go/#hdr-Module_proxy_protocol
+      # TODO: this can take up to 2sec if it's a cache miss on the proxy. Might be able
+      # to scrape the webpage or wait for an API for a faster fetch here.
+
+      version = latest_version_number(name) if version.nil?
+      return if version.nil?
+
+      contents = get_raw("#{PROXY_BASE_URL}/#{encode_for_proxy(name)}/@v/#{encode_for_proxy(version)}.mod")
+
+      Rails.logger.info "[Unable to fetch go.mod contents] name=#{name}" if contents.blank?
+
+      contents
+    end
```

Comment:
If this is only being used by `fetch_mod` there's no need for them to be separate methods. `fetch_mod` can do the actual fetch, then return the `GoMod` object.